keps/sig-network/0031-20181017-kube-proxy-services-optional.md

```diff
@@ -0,0 +1,117 @@
+---
+kep-number: 0
+title: Make kube-proxy service abstraction optional
+authors:
+  - "@bradhoekstra"
+owning-sig: sig-network
+participating-sigs:
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: "@bradhoekstra"
+creation-date: 2018-10-17
+last-updated: 2018-10-17
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Make kube-proxy service abstraction optional
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+      * [Story 1](#story-1)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Design](#design)
+      * [Testing](#testing)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+
+## Summary
+
+In a cluster that has a service mesh a lot of the work being done by kube-proxy is redundant and wasted.
+Specifically, services that are only reached via other services in the mesh will never use the service abstraction implemented by kube-proxy in iptables (or ipvs).
+By informing the kube-proxy of this, we can lighten the work it is doing and the burden on its proxy backend.
+
+## Motivation
+
+The motivation for the enhancement is to allow higher scalability in large clusters with lots of services that are making use of a service mesh.
+
+### Goals
+
+The goal is to reduce the load on:
+* The apiserver sending all services and endpoints to all kube-proxy pods
+* The kube-proxy having to deserialize and process all services and endpoints
+* The backend system (e.g. iptables) for whichever proxy mode kube-proxy is using
+
+### Non-Goals
+
+* Making sure the service is still routable via the service mesh
+* Preserving any kube-proxy functionality for any intentionally disabled Service, including but not limited to: externalIPs, external LB routing, nodePorts, externalTrafficPolicy, healthCheckNodePort, UDP, SCTP
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator, operating a cluster using a service mesh I want to be able to disable the kube-proxy service implementation for services in that mesh to reduce overall load on the whole cluster
+
+### Implementation Details/Notes/Constraints
+
+#### Overview
+
+It is important for overall scalability that kube-proxy does not receive data for Service/Endpoints objects that it is not going to affect. This can reduce load on the apiserver, networking, and kube-proxy itself by never receiving the updates in the first place.
+
+The proposal is to make this feature available by annotating the Service object with this label: `kube-proxy.kubernetes.io/disabled=true`. The associated Endpoints object will automatically inherit that label from the Service object as well.
```

Comment:
What does Endpoints need with that label?  for kube-proxy, Service should be enough, no?

Reply:
kube-proxy watches for both Service and Endpoints objects. It should not receive updates for Endpoints objects that are associated with 'disabled' Service objects, since that traffic/processing would also be wasted.

Reply:
How about adding the labels later after service creation? Should kube-proxy discard the service's data and reset iptables rules/LBs and so on?

Reply:
Yes. Will the proposed implementation, this will happen automatically as the informer will report the Service as deleted when the label gets added.

The reverse is also true, if the label is removed then kube-proxy should add rules for the service. This part is also handled automatically by the filtered informer.

Reply:
If we don't use kube-proxy in the key, it should be clear that this is a hint and not a directive otherwise we run into the issue @smarterclayton mentioned before, which is whether or not it is required for alternate implementations to implement this feature.

I do like the idea of this potentially being used by other implementations, in which case they probably shouldn't have to hard code 'kube-proxy' labels into their code.

A more generic name could be:
`service.kubernetes.io/route-exclusively-via-service-mesh=true`